### PR TITLE
Draft component for Underline Nav

### PR DESCRIPTION
**Pull request by pksjce**
Describe your changes here.

Most of the work for the UnderlineNav component was done in https://github.com/primer/react/pull/2017

The deviation from the original component is pretty high. So decided to go with a new draft component and will follow deprecation process for the original component.

Closes https://github.com/github/primer/issues/892

### Screenshots

TBD

### Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation
- [x] Tested in Chrome
- [x] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

File: src/UnderlineNav/UnderlineNavLink.tsx
Comment (colebemis) [inline src/UnderlineNav/UnderlineNavLink.tsx:81]: Is `selected` a valid prop to pass to `Box`?

File: src/UnderlineNav/UnderlineNavLink.tsx
Comment (colebemis) [inline src/UnderlineNav/UnderlineNavLink.tsx:70]: If you're passing this object to the `sx` prop, you don't need to use the `get()` function:

```suggestion
      paddingX: 2,
      paddingY: 3,
      marginRight: 3,
      fontSize: 1,
```

File: src/UnderlineNav/UnderlineNavLink.tsx
Comment (colebemis) [inline src/UnderlineNav/UnderlineNavLink.tsx:34]: Should accept an `aria-current` prop instead of `selected`? That would be more consistent with NavList: https://primer.style/react/NavList#navlistitem

File: src/UnderlineNav/UnderlineNav.tsx
Comment (mperrotti) [inline src/UnderlineNav/UnderlineNav.tsx:17]: Non-blocking nitpick: Could we remove these comments since they're done?

File: src/UnderlineNav/UnderlineNav.tsx
Comment (mperrotti) [inline src/UnderlineNav/UnderlineNav.tsx:40]: Can't we get the dimensions from `useResizeObserver`? This would help us avoid the reflow that gets triggered when you call `.getBoundingClientRect()`

File: src/UnderlineNav/UnderlineNavLink.tsx
Comment (mperrotti) [inline src/UnderlineNav/UnderlineNavLink.tsx:57]: Could we use a token instead of `'8px'`?

**Review by mperrotti — COMMENTED** [review]
Two more observations:
- `src/UnderlineNav/types.ts` is empty. We could remove it, or move some types to it
- It would be nice to add some tests now so we can add new tests as we add new features instead of writing a whole bunch of tests at once

File: src/UnderlineNav/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav/UnderlineNav.tsx:40]: Ah! this is good catch. I'll try to change the impl to use it. I implemented this part first and then wrapped it with the observer callback. So I didn't think of using it.

Though for first pass I'd still need this. So I'll do a conditional of if I have resizeObserver props, I'll use them, if not no.

File: src/UnderlineNav/UnderlineNavLink.tsx
Comment (pksjce) [reply src/UnderlineNav/UnderlineNavLink.tsx:85]: Yep! overhauling the selected mechanism to be more functional.

File: src/UnderlineNav/UnderlineNavLink.tsx
Comment (pksjce) [reply src/UnderlineNav/UnderlineNavLink.tsx:74]: Thanks! will do this.

File: src/UnderlineNav/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav/UnderlineNav.tsx:17]: Sure! Had it there for personal accountability. Will be deleting these lines.

**Comment by mperrotti** [comment]
👋  Assigning myself to this because I volunteered to be your PR buddy on https://github.com/github/primer/issues/892 in this today's Primer React planning meeting

File: src/UnderlineNav/UnderlineNav.tsx
Comment (mperrotti) [reply src/UnderlineNav/UnderlineNav.tsx]: Any luck on this?

**Review by mperrotti — COMMENTED** [review]
I had a few more pieces of feedback on the big stuff. I could get more detailed now, or hold off until tests are written and this is closer to being ready to merge. Just let me know which you prefer. 🙂

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.tsx:42]: yes! I updated the `useResizeObserver` but somehow missed updating here. My bad!

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.tsx:138]: On the first render, we need this. Or you don't think so? My testing showed that.

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.tsx:24]: Yes! Adding `overflow` and other props in controls would be good.

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.tsx:97]: Yeah! thanks for the feedback. I'm really not sure why this reverse align feature needs to exist. I only added it because prev component has it. 
Also should that "More" dropdown reverse direction or not? Do you have any design input here?

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.tsx:33]: When the window is smaller, we are not able to show all the items. I calculate and convert the overflowing items into "actions" and put them in an `ActionMenu` dropdown currently labelled "More"

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.tsx:141]: Orly!? Lemme debug why?

Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.tsx:141]: Oh no! it does do that. Maybe `requestAnimationFrame` is not the way to go here. Will need to think some more.

**Comment by pksjce** [comment]
> I had a few more pieces of feedback on the big stuff. I could get more detailed now, or hold off until tests are written and this is closer to being ready to merge. Just let me know which you prefer.

@mperrotti Would love any feedback whenever you have the time to give it. Thanks a lot for your help with this

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (mperrotti) [reply src/UnderlineNav2/UnderlineNav.tsx:141]: My guess is that some function in there needs to be in a `useCallback`

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (mperrotti) [reply src/UnderlineNav2/UnderlineNav.tsx:138]: I commented it out, and it behaved as expected for me.

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (mperrotti) [reply src/UnderlineNav2/UnderlineNav.tsx:97]: I think the "More" dropdown should be next to the last tab like it is in my screenshot.

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.tsx]: Yes! My fix was the same. The callback to useResizeObserver didn't have any constraints. It's fixed now

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.tsx]: oh yeah 🤷‍♀️ its true! I'm finding the same.

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (mperrotti) [inline src/UnderlineNav2/UnderlineNav.tsx:31]: Why do we have to use `forEach` instead of a [for...of loop](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of)?

File: docs/content/UnderlineNav2.mdx
Comment (mperrotti) [inline docs/content/UnderlineNav2.mdx:48]: I think this needs to be updated with the new `onSelect` prop info

File: src/UnderlineNav2/UnderlineNav.stories.tsx
Comment (mperrotti) [inline src/UnderlineNav2/UnderlineNav.stories.tsx:44]: I didn't even know `.play` was a thing! Do we want to keep it in here, or was this only for debugging?

File: src/UnderlineNav2/UnderlineNav.stories.tsx
Comment (mperrotti) [inline src/UnderlineNav2/UnderlineNav.stories.tsx:67]: I don't think we need this story since we can change the alignment in the controls.

File: docs/content/UnderlineNav2.mdx
Comment (mperrotti) [inline docs/content/UnderlineNav2.mdx:1]: I think this file should go in `docs/content/drafts/`

File: docs/content/UnderlineNav2.mdx
Comment (mperrotti) [inline docs/content/UnderlineNav2.mdx:2]: Since we already have an "UnderlineNav" that this is replacing, I think this title needs to be something like "UnderlineNav v2" to differentiate it.

**Review by mperrotti — COMMENTED** [review]
Looking great! Just a few minor comments/questions.

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (mperrotti) [inline src/UnderlineNav2/UnderlineNav.tsx:51]: Why did we choose 1/2 of the width as the cutoff? I think we could pick a higher threshold, or calculate based on the width of UnderlineNav minus the width of the dropdown button.

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.tsx:31]: I just found it convenient to get the `index` in this case. But I guess you are right. This is an eslint disable I can remove from code.

File: docs/content/UnderlineNav2.mdx
Comment (pksjce) [reply docs/content/UnderlineNav2.mdx:48]: So the `onSelect` is on `UnderlineNavLink`, and the `afterSelect` is on `UnderlineNav`. I found this pattern in `ActionList`. Do you think we need to change it?

File: src/UnderlineNav2/UnderlineNav.stories.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.stories.tsx:44]: These are new interaction tests! I've some minimal documentation in contribution docs about it. I'm working on an ADR for detailed guidelines too. It was my passion week project. They are very similar to integration tests.

File: src/UnderlineNav2/UnderlineNav.stories.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.stories.tsx:67]: Well.. I'd put it in docs. so I thought I'd have a story for it as well.

File: docs/content/UnderlineNav2.mdx
Comment (pksjce) [reply docs/content/UnderlineNav2.mdx:1]: oops! My bad. Will fix this.,

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.tsx:51]: Yeah, this calculation really didn't evolve with feedback. But I feel we could tweak it after merging the draft? I'm figuring out how to write tests for this. Then we can experiment with different cut-offs?

**Comment by maximedegreve** [comment]
Could we also add an example to the docs for usage with the [<CounterLabel />](https://primer.style/react/CounterLabel) component since this is common use case.

File: docs/content/UnderlineNav2.mdx
Comment (mperrotti) [reply docs/content/UnderlineNav2.mdx:48]: We can keep afterSelect - just update the `UnderlineNavProps` type. Currently, I'm seeing `onSelect` and not `afterSelect`

File: src/UnderlineNav2/UnderlineNav.stories.tsx
Comment (mperrotti) [reply src/UnderlineNav2/UnderlineNav.stories.tsx:44]: Ahh very cool! I'm ok with keeping it if you feel like it's ready. 

It would be nice to get that ADR PR opened sooner than later so the rest of the team can understand how to use the new integration tests.

File: src/UnderlineNav2/UnderlineNav.tsx
Comment (pksjce) [reply src/UnderlineNav2/UnderlineNav.tsx]: done!

**Review by mperrotti — APPROVED** [review]

**Pull request merged**
